### ISSUE TITLE
[info] What does CC65 do with the hardware stack? 

### ISSUE THREAD

**Issue by Fabrizio-Caruso** [cc65/cc65#1037]
Is the hardware stack (page 1) underused by CC65?
If I understand well, parameters passing is done through registers + software stack.
Why is the software stack used in this case? I know the hardware stack is tiny but it would be cool to have an option to use it in programs that pass just a couple of parameters beyond the 2 bytes that are usually handled by registers.
It seems to me that the hardware stack is not used a lot. 
I even use the lower part of it to stack data in my programs.
Wouldn't the hardware stack bring more speed compared to the software one if used for parameter passing?

**Comment by mrdudz** [comment]
the problem with doing that is to do it in a generic way, that will still work with non trivial programs. since the return addresses also go to the hardware stack, you'll probably run out of stack space quickly with reasonably deep nested calls

**Comment by Fabrizio-Caruso** [comment]
Yes, I know how small the hardware stack is. My project is a big one with tons of function calls. The hardware stack is underused in my not so trivial case. I am basically allocating variables on the lower part of the stack.
I don't know how complex it would be to have the hardware stack handle parameters as an option similar to static locals. It would help a lot in gaining both speed and memory.

**Comment by oliverschmidt** [comment]
My two cents:
- C programs not written for cc65 "all" require much more stack space than the h/w stack can provide.
- C programs written for cc65 don't make much use of function parameters anyhow. They are never fast, no matter where/how they would be stored.

In general I personally consider cc65 already complex "enough". I don't see that all those special cases like zero page variables defined in C, automatic variables on the h/w stack, <you name it> have their place in cc65. After all the original author of the compiler abandoned the project and none of the current maintainers is able to make significant changes to the compiler. So to be honest the only reason why I don't close feature requests like this right away is pure courtesy.

**Comment by dmsc** [comment]
Another two cents:
- Using the hardware stack for passing parameters on the 6502 adds a lot of code, as it needs trashing of the X register and manually fixing the stack before the return.
- A much better use of the hardware stack would be to store callee saved registers (instead of the software stack currently used). This would mean changing function prologue/epilogue generation, and would be a less intrusive change.
- If you are worried you are "wasting" stack memory, you can modify the linker scripts adding RAM at the low stack addresses, so you can use it for storing static variables or other data.

**Comment by Fabrizio-Caruso** [comment]
@dmsc, Thanks for your explanation! I am using the stack by using an Assembly file with no Assembly in it and with export directives.

**Comment by Fabrizio-Caruso** [comment]
@dmsc Your suggest approach would even work for function calls with no parameters as (I suppose) registers have to be restored in any case.

**Comment by clbr** [comment]
It'd only be useful on a 65816 really, since that cpu added stack addressing "lda 1,s".

**Comment by Fabrizio-Caruso** [comment]
@clbr I see that the 65816 would use the hardware stack even better but I cannot understand why the pushing and popping values into the hardware stack isn't faster than doing the same on a software stack even on the good old 6502.
My lack from understanding comes from little knowledge.

**Comment by clbr** [comment]
A function may access a param many times. It's not just popped once (though you could imagine a prologue that pops all args and puts them to a temp storage...).

Imagine how slow it would be to access, say, the second func param on the hw stack.

**Comment by Fabrizio-Caruso** [comment]
@clbr then I have not understood what you are saying.
Maybe you are not talking about dmsc's second solution where the hardware stack could be used to  store the callee saved registers.
I understand that the function parameters cannot be really optimized with the stack (other than maybe just one byte).

**Comment by Fabrizio-Caruso** [comment]
@clbr, have I missed dome important points? Is the second approach (restore registers) equally not worthwhile?

**Comment by clbr** [comment]
Those are rarely used. Only regbank is callee saved, and most of my code has never generated accesses to that.

**Comment by mrdudz** [comment]
closing this one
